File: rust-rsc-prototype/app/catalog/js/page.js
const dataOrigin = process.env.CATALOG_DATA_ORIGIN || 'http://127.0.0.1:3041'
const benchmarkMode = process.env.CATALOG_BENCHMARK_MODE === '1'
const data = require('../../../catalog/query-data')
import { Suspense } from 'react'
import CatalogControls from '../controls'
const warmCache = new Map()
function wait(delay) {
  return delay > 0
    ? new Promise((resolve) => setTimeout(resolve, Math.min(delay, 5000)))
    : Promise.resolve()
}
async function getJson(path, cacheMode, requestCache) {
  const key = `${dataOrigin}${path}`
  if (cacheMode === 'warm' && warmCache.has(key)) return warmCache.get(key)
  if (cacheMode !== 'uncached' && requestCache.has(key))
    return requestCache.get(key)
  const work = benchmarkMode
    ? fetch(key, {
        cache: 'no-store',
        signal: AbortSignal.timeout(5000),
      }).then(async (response) => {
        if (!response.ok)
          throw new Error(`catalog data returned ${response.status}`)
        return response.json()
      })
    : localJson(path)
  if (cacheMode !== 'uncached') requestCache.set(key, work)
  if (cacheMode === 'warm') warmCache.set(key, work)
  return work
}
async function localJson(path) {
  const url = new URL(path, 'http://catalog.local')
  await wait(Number(url.searchParams.get('delay') || 0))
  return url.pathname === '/categories'
    ? data.categories()
    : data.products(url.searchParams)
}
function value(params, name, fallback) {
  const item = params[name]
  return Array.isArray(item) ? item[0] : item || fallback
}
export default async function JavaScriptCatalog({ searchParams }) {
  const params = await searchParams
  const category = value(params, 'category', 'all'),
    query = value(params, 'q', ''),
    material = value(params, 'material', 'all'),
    sort = value(params, 'sort', 'name'),
    page = value(params, 'page', '1'),
    delay = value(params, 'delay', '')
  const categoryDelay = value(
      params,
      'categoryDelay',
      delay === '' ? '350' : delay
    ),
    productDelay = value(params, 'productDelay', delay === '' ? '1000' : delay)
  const cacheMode = value(params, 'cache', 'uncached')
  if (!['uncached', 'request', 'warm'].includes(cacheMode))
    throw new Error('invalid catalog cache mode')
  const requestCache = new Map()
  const queryString = new URLSearchParams({
    category,
    q: query,
    material,
    sort,
    page,
    delay: productDelay,
  })
  const categories = getJson(
    `/categories?delay=${encodeURIComponent(categoryDelay)}`,
    cacheMode,
    requestCache
  )
  const products = getJson(`/products?${queryString}`, cacheMode, requestCache)
  return (
    <CatalogControls basePath="/catalog/js">
      <div className="catalog-shell" data-catalog="js">
        <header className="catalog-header">
          <a href="/catalog/js" className="catalog-brand">
            RustWorks Supply
          </a>
          <form className="catalog-search">
            <label htmlFor="catalog-q">Search products</label>
            <input id="catalog-q" name="q" defaultValue={query} />
            <button>Search</button>
          </form>
        </header>
        <div className="catalog-grid">
          <Suspense fallback={<CategoriesLoading />}>
            <CategoriesRegion categories={categories} />
          </Suspense>
          <Suspense fallback={<ProductsLoading />}>
            <ProductsRegion
              products={products}
              {...{ query, material, sort, page }}
            />
          </Suspense>
        </div>
      </div>
    </CatalogControls>
  )
}
function CategoriesLoading() {
  return (
    <nav
      className="catalog-sidebar catalog-pagelet"
      aria-label="Categories"
      aria-busy="true"
      data-loading-region="categories"
    >
      <p className="catalog-pagelet-label">Loading categories…</p>
      <div className="catalog-skeleton-lines" aria-hidden="true">
        <i />
        <i />
        <i />
        <i />
      </div>
    </nav>
  )
}
function ProductsLoading() {
  return (
    <main
      className="catalog-pagelet"
      aria-label="Products"
      aria-busy="true"
      data-loading-region="products"
    >
      <p className="catalog-pagelet-label">Loading product pagelet…</p>
      <div className="catalog-skeleton-table" aria-hidden="true">
        <i />
        <i />
        <i />
        <i />
        <i />
      </div>
    </main>
  )
}
async function CategoriesRegion({ categories }) {
  const resolved = await categories
  return (
    <nav aria-label="Categories" className="catalog-sidebar">
      <h2>Categories</h2>
      <a href="?category=all" data-catalog-navigation>
        All products <span>720</span>
      </a>
      {resolved.map((item) => (
        <a key={item.id} href={`?category=${item.id}`} data-catalog-navigation>
          {item.name} <span>{item.count}</span>
        </a>
      ))}
    </nav>
  )
}
async function ProductsRegion({ products, query, material, sort, page }) {
  const result = await products
  return (
    <main>
      <form className="catalog-filters">
        <input type="hidden" name="q" value={query} readOnly />
        <input type="hidden" name="productDelay" value="1000" readOnly />
        <label>
          Material{' '}
          <select name="material" defaultValue={material}>
            <option value="all">All</option>
            <option>Zinc Steel</option>
            <option>Stainless Steel</option>
            <option>Aluminum</option>
            <option>Brass</option>
          </select>
        </label>
        <label>
          Sort{' '}
          <select name="sort" defaultValue={sort}>
            <option value="name">Name</option>
            <option value="price">Price</option>
          </select>
        </label>
        <button>Apply</button>
      </form>
      <p className="catalog-summary">
        {result.total} products · page {page}
      </p>
      <table className="catalog-table">
        <thead>
          <tr>
            <th>Part</th>
            <th>Description</th>
            <th>Material</th>
            <th>Size</th>
            <th>Available</th>
            <th>Price</th>
          </tr>
        </thead>
        <tbody>
          {result.products.map((product) => (
            <tr key={product.id} data-product-id={product.id}>
              <td>
                <a
                  href={`/catalog/js/product/${product.id}`}
                  data-catalog-navigation
                >
                  {product.id}
                </a>
              </td>
              <td>
                {product.name}
                <small>{product.finish}</small>
              </td>
              <td>{product.material}</td>
              <td>{product.size}</td>
              <td>{product.available}</td>
              <td>${(product.priceCents / 100).toFixed(2)}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </main>
  )
}

File: rust-rsc-prototype/app/catalog/rust/page.rs
use next_rsc::{Node, PageProps, ParamValue, RenderResult, element};

const CATEGORIES: [(&str, &str); 6] = [
    ("fasteners", "Fasteners"),
    ("electrical", "Electrical"),
    ("plumbing", "Plumbing"),
    ("material-handling", "Material Handling"),
    ("safety", "Safety"),
    ("machining", "Machining"),
];
const MATERIALS: [&str; 4] = ["Zinc Steel", "Stainless Steel", "Aluminum", "Brass"];
const FINISHES: [&str; 4] = ["Plain", "Black Oxide", "Galvanized", "Anodized"];

struct Product {
    id: String,
    category: &'static str,
    name: String,
    material: &'static str,
    finish: &'static str,
    size: String,
    price_cents: usize,
    available: usize,
}

fn query(props: &PageProps, name: &str, fallback: &str) -> String {
    match props.search_params.get(name) {
        Some(ParamValue::String(value)) => value.clone(),
        Some(ParamValue::Strings(values)) => values
            .first()
            .cloned()
            .unwrap_or_else(|| fallback.to_owned()),
        None => fallback.to_owned(),
    }
}

fn link(href: String, label: impl Into<String>) -> Node {
    element("a", [Node::text(label.into())])
        .prop("href", href)
        .prop("data-catalog-navigation", true)
}

fn category_link(href: String, label: impl Into<String>, count: usize) -> Node {
    element(
        "a",
        [
            Node::text(format!("{} ", label.into())),
            element("span", [Node::text(count.to_string())]),
        ],
    )
    .prop("href", href)
    .prop("data-catalog-navigation", true)
}

fn product(index: usize) -> Product {
    let (category, category_name) = CATEGORIES[index % CATEGORIES.len()];
    Product {
        id: format!("RC-{:05}", index + 1),
        category,
        name: format!("{category_name} {:03}", (index % 120) + 1),
        material: MATERIALS[index % MATERIALS.len()],
        finish: FINISHES[(index / 3) % FINISHES.len()],
        size: format!("{} mm", (index % 24) + 1),
        price_cents: 175 + ((index * 137) % 18500),
        available: 8 + ((index * 29) % 940),
    }
}

pub fn render(props: PageProps) -> RenderResult {
    let category = query(&props, "category", "all");
    let search = query(&props, "q", "");
    let material = query(&props, "material", "all");
    let sort = query(&props, "sort", "name");
    let page = query(&props, "page", "1")
        .parse::<usize>()
        .unwrap_or(1)
        .max(1);
    let normalized_search = search.to_lowercase();
    let mut products = (0..720)
        .map(product)
        .filter(|product| {
            (category == "all" || product.category == category)
                && (material == "all" || product.material == material)
                && (normalized_search.is_empty()
                    || format!("{} {} {}", product.id, product.name, product.material)
                        .to_lowercase()
                        .contains(&normalized_search))
        })
        .collect::<Vec<_>>();
    if sort == "price" {
        products.sort_by_key(|product| product.price_cents);
    } else {
        products.sort_by(|left, right| {
            left.name
                .cmp(&right.name)
                .then_with(|| left.id.cmp(&right.id))
        });
    }
    let total = products.len();
    let products = products
        .into_iter()
        .skip((page - 1) * 24)
        .take(24)
        .collect::<Vec<_>>();

    let category_links = std::iter::once(category_link(
        "?category=all".to_owned(),
        "All products",
        720,
    ))
    .chain(
        CATEGORIES
            .into_iter()
            .map(|(id, name)| category_link(format!("?category={id}"), name, 120)),
    );
    let sidebar = element(
        "nav",
        [
            element("h2", [Node::text("Categories")]),
            Node::fragment(category_links),
        ],
    )
    .prop("aria-label", "Categories")
    .prop("className", "catalog-sidebar");

    let rows = products.into_iter().map(|product| {
        let id = product.id.clone();
        element(
            "tr",
            [
                element(
                    "td",
                    [link(format!("/catalog/rust/product/{id}"), id.clone())],
                ),
                element(
                    "td",
                    [
                        Node::text(product.name),
                        element("small", [Node::text(product.finish)]),
                    ],
                ),
                element("td", [Node::text(product.material)]),
                element("td", [Node::text(product.size)]),
                element("td", [Node::text(product.available.to_string())]),
                element(
                    "td",
                    [Node::text(format!(
                        "${:.2}",
                        product.price_cents as f64 / 100.0
                    ))],
                ),
            ],
        )
        .prop("data-product-id", id)
    });
    let option =
        |value: &str, label: &str| element("option", [Node::text(label)]).prop("value", value);
    let filters = element(
        "form",
        [
            element("input", [])
                .prop("type", "hidden")
                .prop("name", "q")
                .prop("defaultValue", search.clone()),
            element("input", [])
                .prop("type", "hidden")
                .prop("name", "productDelay")
                .prop("defaultValue", "1000"),
            element(
                "label",
                [
                    Node::text("Material "),
                    element(
                        "select",
                        [
                            option("all", "All"),
                            option("Zinc Steel", "Zinc Steel"),
                            option("Stainless Steel", "Stainless Steel"),
                            option("Aluminum", "Aluminum"),
                            option("Brass", "Brass"),
                        ],
                    )
                    .prop("name", "material")
                    .prop("defaultValue", material),
                ],
            ),
            element(
                "label",
                [
                    Node::text("Sort "),
                    element("select", [option("name", "Name"), option("price", "Price")])
                        .prop("name", "sort")
                        .prop("defaultValue", sort),
                ],
            ),
            element("button", [Node::text("Apply")]),
        ],
    )
    .prop("className", "catalog-filters");
    let headings = [
        "Part",
        "Description",
        "Material",
        "Size",
        "Available",
        "Price",
    ]
    .into_iter()
    .map(|name| element("th", [Node::text(name)]));
    let content = element(
        "main",
        [
            filters,
            element("p", [Node::text(format!("{total} products · page {page}"))])
                .prop("className", "catalog-summary"),
            element(
                "table",
                [
                    element("thead", [element("tr", headings)]),
                    element("tbody", rows),
                ],
            )
            .prop("className", "catalog-table"),
        ],
    );
    let search_form = element(
        "form",
        [
            element("label", [Node::text("Search products")]).prop("htmlFor", "catalog-q"),
            element("input", [])
                .prop("id", "catalog-q")
                .prop("name", "q")
                .prop("defaultValue", search),
            element("button", [Node::text("Search")]),
        ],
    )
    .prop("className", "catalog-search");
    Ok(element(
        "div",
        [
            element(
                "header",
                [
                    link("/catalog/rust".to_owned(), "RustWorks Supply")
                        .prop("className", "catalog-brand"),
                    search_form,
                ],
            )
            .prop("className", "catalog-header"),
            element("div", [sidebar, content]).prop("className", "catalog-grid"),
        ],
    )
    .prop("className", "catalog-shell")
    .prop("data-catalog", "rust"))
}
